### ISSUE TITLE
Make BSP import more robust

### ISSUE THREAD

**Issue by nafg** [com-lihaoyi/mill#3011]
(Previously posted at https://discord.com/channels/632150470000902164/940067748103487558/1207105622366228480)

With a project imported from Mill into IntelliJ via BSP, I regularly have scenarios like the following:

1. I update dependencies and sync (reimport the project)
2. It fails with something cryptic like this: ![image](https://github.com/com-lihaoyi/mill/assets/98384/11735e81-ea95-4318-8db6-852ac823bbbc)
3. Now the entire project is red, so the IDE is very unheplful
4. I check mill-bsp.stderr and see something like
```
buildTargetScalacOptions caught exception: java.lang.Exception: Failure during task evaluation: app_common.test.compile Compilation failed
java.lang.Exception: Failure during task evaluation: app_common.test.compile Compilation failed
    at mill.eval.Evaluator.$anonfun$evalOrThrow$default$1$1(Evaluator.scala:42)
    at mill.eval.EvaluatorImpl$EvalOrThrow.apply(EvaluatorImpl.scala:67)
```
In short, unless there are no compile errors anywhere when I sync the build it breaks the whole IDE with no easy to find reason

This does not happen with SBT BSP import.

Suggestions:
A. If importing fails, it shouldn't make things worse in the IDE than before
B. A module failing to compile is an an exception, it should be a case the BSP code thinks about
C. Ideally it would be able to give IDE whatever information it can
D. Errors shouldn't be hidden somewhere inside a ginormous log file, they should be returned to the IDE

**Comment by nafg** [comment]
Another thing is that often after reimporting -Xsource:3 syntax like `*` imports are red until I do Repair IDE and get past Reopen Project (I confess, I'm not sure if I tried going straight to manually closing and reopening the project).

**Comment by lihaoyi** [comment]
I've hit this too. We should not need to compile everything to generate the IDE project

I vaguely remember fixing a bug like this before, but maybe I'm mistaken or it regressed

**Comment by lihaoyi** [comment]
In my case I suspect it's because I'm doing unusual things, e.g. compiling compiler plugins as part of the build, or depending on output `.jar` targets as part of `unmanagedClasspath`. Doing that, compiling stuff when generating the IDE project is unavoidable

@nafg is there any chance you are doing something similar?

**Comment by lefou** [comment]
Would be good to have a reproducer for that.

**Comment by lihaoyi** [comment]
Looking at Nagtoli's screenshots, it seens that the compile is failing when computing `scalacOptions`, which does hint at a compiler plugin being compiled and passed as a `-Xplugin-classpath` flag. That would be the same root cause as the case I hit

The other quality-of-life improvements we should still do though, or get the IntelliJ folks to do. We probably shouldn't crash the entire project's IDE setup just because one module failed to compile, and when things do crash we should show the error nicely rather than having to dig through a huge log file

**Comment by lihaoyi** [comment]
A simple repro would be a two module project with one module's `unmanagedClasspath` being set to the second module's `compile()` output. It's a bit contrived, but it should demonstrate the IDE import crashing if the second module fails to compile

**Comment by lefou** [comment]
I guess, it's in the nature of the BSP protocol, that we need to fail in cases we hit some failure along the way. The granularity is determined the by request, and the return type is all we can use to communicate result or failure. In BSP, besides compilation and some other requests, where compile result and failure is part of the result or notification, in scalacOptions there is no failure anticipated. Without having analyzed what these other tools do, I'd assume they either don't hit an error or are not using the protocol specified error channels. We simply don't have control over the granularity the client requests some data. Typically, clients send a single request for _all_ build targets (= Mill modules).

What options do we have, if we hit an error in `scalacOptions`? Is just returning some wrong result (like an empty `Seq`) combined with some loosely coupled log error message sent to the client, ok? I think, the BSP spec says nothing about it.

It may be a general protocol design flaw. Currently, you can either send many small request, which might impose some overhead, or you send large requests, which take long and if they fail provide nothing. Instead, we should consider either more granular error results, e.g. per build target, or stream the results as they come, as we can do partially with compilation messages.

**Comment by nafg** [comment]
Somehow I don't have these issues with sbt or bleep

Sent with Shortwave <https://www.shortwave.com?utm_medium=email&utm_content=signature&utm_source=bmFmdG9saWd1Z0BnbWFpbC5jb20=>

**Comment by lefou** [comment]
> ```
> buildTargetScalacOptions caught exception: java.lang.Exception: Failure during task evaluation: app_common.test.compile Compilation failed
> java.lang.Exception: Failure during task evaluation: app_common.test.compile Compilation failed
>     at mill.eval.Evaluator.$anonfun$evalOrThrow$default$1$1(Evaluator.scala:42)
>     at mill.eval.EvaluatorImpl$EvalOrThrow.apply(EvaluatorImpl.scala:67)
> ```

@nafg Can you run `mill app_common.test.compile`? Can you see, why it failed? If it is not failing, can you see in the logs why/how it failed when ran as part of `BSP/install`?

**Comment by nafg** [comment]
It was an ordinary compilation error

Sent with Shortwave <https://www.shortwave.com?utm_medium=email&utm_content=signature&utm_source=bmFmdG9saWd1Z0BnbWFpbC5jb20=>

**Comment by lefou** [comment]
> Another thing is that often after reimporting -Xsource:3 syntax like `*` imports are red until I do Repair IDE and get past Reopen Project (I confess, I'm not sure if I tried going straight to manually closing and reopening the project).

@nafg This is unrelated to the initial issue. If you think this is an issue in Mill, please open a new issue.

**Comment by lihaoyi** [comment]
I think for `scalacOptions` in particular, falling back to an empty `Seq()` if evaluation fails is very reasonable. That would catch the most common case this appears - of compiling compiler plugins as part of the build - and would be no worse than the status quo of "nothing works, at all"

**Comment by nafg** [comment]
> In my case I suspect it's because I'm doing unusual things, e.g. compiling compiler plugins as part of the build, or depending on output `.jar` targets as part of `unmanagedClasspath`. Doing that, compiling stuff when generating the IDE project is unavoidable
> 
> @nafg is there any chance you are doing something similar?

I don't think so. I do customize localClasspath somewhat conditionally but still:

```
❯ mill inspect app_common.localClasspath
[1/1] inspect 
app_common.localClasspath(build.sc:77)
    The *output* classfiles/resources from this module, used for execution,
    excluding upstream modules and third-party dependencies

Inputs:
    app_common.localCompileClasspath
    app_common.resources
    app_common.compile
    app_common.webapp
    app_common.generatedBuildInfo

❯ mill inspect app_common.generatedBuildInfo 
[1/1] inspect 
app_common.generatedBuildInfo(build.sc:38)

Inputs:

```

It's based on

```scala
  override def localClasspath =
    if (scalaJsAsAsset)
      T(super.localClasspath() ++ Seq(generatedBuildInfo(), app_js.getScalaJs()))
    else
      T(super.localClasspath() :+ generatedBuildInfo())
```

where

```scala
  def scalaJsAsAsset = insideCI
```

where

```scala
def insideCI = sys.env.get("CI").exists(Set("", "true", "1"))
```

Also, `scalacOptions` has a dependency on `scalaVersion`.

**Comment by nafg** [comment]
> > Another thing is that often after reimporting -Xsource:3 syntax like `*` imports are red until I do Repair IDE and get past Reopen Project (I confess, I'm not sure if I tried going straight to manually closing and reopening the project).
> 
> @nafg This is unrelated to the initial issue. If you think this is an issue in Mill, please open a new issue.

What do I know, but to me "errors in importing results in everything red" and "successful importing results in spurious red code" (when with sbt importing can fail and doesn't have either issue) seems like part of the same problem (which are unrelated to "why are compile errors causing import to fail" which is not the main issue.)

**Comment by lefou** [comment]
Opened discussion: https://github.com/build-server-protocol/build-server-protocol/discussions/655

**Comment by adpi2** [comment]
> What options do we have, if we hit an error in scalacOptions? Is just returning some wrong result (like an empty Seq) combined with some loosely coupled log error message sent to the client, ok? I think, the BSP spec says nothing about it.

We had the same issue in sbt and I solved it (https://github.com/sbt/sbt/pull/6609) by sending back a response with missing build targets. For instance if the clients want the `scalacOptions` of target `a`, `b` and `c`, and if there is a failure for `b`, sbt will send back a response with the `scalacOptions` of `a` and `c`. This is not implemented only for `scalacOptions` but for all multi-target requests.

I think I recall that Bloop has the same behavior and that's why I decided to implement it this way in sbt. On the client side Metals should handle such incomplete response quite well, for IntelliJ I am not sure.

**Comment by lefou** [comment]
@adpi2 Thank you for this response. This sounds like a nice generic approach. I think, I'm going to implement that in Mill as well. The only downside is, that the BSP client has no idea why the response was incomplete. We could push some informal log message to the client, though.

**Comment by adpi2** [comment]
> The only downside is, that the BSP client has no idea why the response was incomplete.

Maybe it is time to implement https://github.com/build-server-protocol/build-server-protocol/issues/204 then.

**Comment by lefou** [comment]
I've now gone with the approach proposed by @adpi2. If I encounter failure for some requested target identifiers, I just does not include them in the otherwise successful result. In parallel, I send a log message with ERROR level to the BSP client.

@nafg If you could test PR #3022, and report back, whether that matches your expectations, that would be great.

**Comment by nafg** [comment]
How do I test it?

**Comment by lefou** [comment]
> How do I test it?

Once, our CI has some free cycles, I'll start a snapshot build. If you're eager, you can also compile the PR or the `main` branch with `mill installLocalCache` and use the version (`mill show main.publishVersion`) in your `.mill-version`.

**Comment by nafg** [comment]
Just tested it and it seems better already!
